docs/proposals/20191016-clusterctl-redesign.md

```diff
@@ -0,0 +1,265 @@
+
+## Proposal
+
+In this section we will outline a high level overview of the proposed tool and its associated user stories.
+
+### Preconditions
+
+Prior to running clusterctl it is required that the operator has a valid KUBECONFIG + context to a running management cluster, following the same rules as kubectl.
+
+### User Stories
+
+#### Initial Deployment
+
+- As a Kubernetes operator, I’d like to have a simple way for clusterctl to install Cluster API providers into a management cluster, using a very limited set of resources and prerequisites, ideally from a laptop with something I can install with just a couple of terminal commands.
+- As a Kubernetes operator, I’d like to have the ability to provision clusters on different providers, using the same management cluster.
+- As a Kubernetes operator I would like to have a consistent user experience to configure and deploy clusters across cloud providers.
+
+#### Day Two Operations “Lifecycle Management”
+
+- As a Kubernetes operator I would like to be able to install new Cluster API providers into a management cluster.
+- As a Kubernetes operator I would like to be able to upgrade Cluster API components (including CRDs, controllers, etc).  
+- As a Kubernetes operator I would like to have a simple user experience to cleanly remove the Cluster API objects from a management cluster.
+
+#### Target Cluster Pivot Management
+
+- As a Kubernetes operator I would like to pivot Cluster API components from a management cluster to a target cluster.
+
+#### Provider Enablement
+
+- As a Cluster API provider developer, I would like to use my current implementation, or a new version, with clusterctl without recompiling.  
+
+### Implementation Details/Notes/Constraints
+
+> Portions of clusterctl may require changes to the Cluster API data model and/or the definition of new conventions for the providers in order to enable sets of user stories listed above, therefore a roadmap and feature deliverables will need to be coordinated over time.
+
+As of this writing, we envision clusterctl’s component model to look similar to the diagram below:
+
+![components](images/clusterctl-redesign/components.png)
+
+Clusterctl consists of a single binary artifact implementing a CLI; the same features exposed by the CLI will be made available to other management tools via a client library.
+
+During provider installation, clusterctl internal pkg/lib is expected to access provider repositories and read the yaml file specifying all the provider components to be installed in the management cluster.
+
+If the provider repository contains deployment templates, or flavors, clusterctl could be used to generate yaml for creating new workload clusters.  
+
+Behind the scenes clusterctl will apply labels to any and all providers’ components and use a set of custom objects for keeping track of some metadata about installed providers, like e.g. the current version or the namespace where the components are installed. 
```

Comment:
Doesn't this seem like an overlap with other sort of installation tools that exist or kustomize?

Reply:
behind the scene it uses kustomize :-;
the value added is that clusterctl enforce a consistent set of labels that can be used for implementing additional lifecycle actions like e.g upgrade, pivot, delete